### PR TITLE
fixes bug 1085530 - store uploaded symbols to S3

### PR DESCRIPTION
**Pull request by peterbe**
@rhelmer r? cc @twobraids 

So basically, we stop storing the uploaded file. Instead, we immediately unpack it and send it straight into S3. All within one big request.

### CODE REVIEW

**Comment by rhelmer** [comment]
This looks/works great, we have been chatting in IRC trying to work out what happens if an in-progress upload fails.

Right now any upload failure to S3 causes everything to fail, which means the client gets an error and should retry (good). However it means we also don't record anything in the database - I think we should write a row in the database for each file that's uploaded successfully to S3.

**Comment by peterbe** [comment]
For the record, we discussed the risk of lack of atomicity. 
The django side is fully atomic. Meaning if the upload (half-way through) fails because the S3 network connection blows up nothing is recorded in our database and the client gets a 500 error. 
That would mean we'd potentially get some symbols uploaded to S3 that we don't have a record of. 

Some solutions:
1) We currently make a listing of the zip file as a blog of text and later proceed to upload each file. 
We could instead do something like this:
```python
# pseudo code
success = []
try:
    for file in zipfile:
        s3.upload(file)
        success.append(file)
    failed = False
except S3NetworkError:
    failed = True
SymbolsUpload.objects.create(
    user=request.user,
    files='\n'.join(success),
    failed=failed,
    size=zipfile.size,
    name=zipfile.name
)
```

2) We could delete what we've uploaded if it fails. But only delete what wasn't already there. 
This will almost always be tricky if we override a file in S3. 

Also, a risk I can see is that our local postgres connection might fail DURING the request. That means we could be uploading X% symbols but when it's upload 100% our DB connection to postgres might have broken and we can't write down what was upload. 
A solution to that would be to do something like this:
```python
upload = SymbolsUpload.objects.create(
    user=request.user,
    size=zipfile.size,
    name=zipfile.name,
    content=''
)
for file in zipfile:
    s3.bucket.upload(file)
    upload.content += '%s\n' % file
    upload.save()
```

File: webapp-django/crashstats/symbols/views.py [hunk 1]
```diff
@@ -32,6 +36,46 @@ def check_symbols_archive_content(content, header_length=2):
...
+        bucket = conn.get_bucket(settings.SYMBOLS_BUCKET_NAME)
+    except boto.exception.S3ResponseError:
+        bucket = conn.create_bucket(
+            settings.SYMBOLS_BUCKET_NAME,
+            settings.SYMBOLS_BUCKET_LOCATION
+        )
+
+    total_uploaded = 0
+    for member in iterator:
+        key = bucket.new_key(os.path.join(
```
Comment (rhelmer) [inline webapp-django/crashstats/symbols/views.py]: We expect to see a lot of attempts to upload symbols that are already present in S3. We should do a HEAD here and compare the size of the remote and local files, and not bother to upload if it's already present (```key.get_key()``` can do this, see http://boto.readthedocs.org/en/latest/ref/s3.html#boto.s3.bucket.Bucket.get_key)

**Comment by rhelmer** [comment]
@peterbe I think the 3rd solution is the way to go - I don't think the large number of updates will be a big deal since they are all on the same connection and in the same transaction.

**Comment by peterbe** [comment]
We discussed, here's what we will do:
1. For each file we upload/or not we write a link (full S3 URL that contains the bucket name, key, etc) in our big text field 
2. Just before uploading, we check if the key already exists and if it has the same size. E.g. (`boto.get_key()`) If the file does EXISTS we don't bother uploading again. But if the size is different, we upload again.
3. Each file in the zip file will go into the text field but we'll prefix each line with a `=` if the file wasn't re-uploading and `+` if it was uploaded.

**Comment by peterbe** [comment]
@rhelmer time for another review. 
So, I implemented what was mentioned in the comment above **except** instead of storing the absolute URL in the text field I decided on my own format:
```
# when the file didn't exist or the size was different
+BUCKET_NAME,PREFIX/FILENAME
# when the upload was cancelled because it wasn't sufficiently different
=BUCKET_NAME,PREFIX/FILENAME
```

For example:
screenshot 2015-02-05 10 48 09

The reason I didn't use a URL was because then you have to decide on an expiration time and the URL will contain a key. 
Also, suppose the filename contains something that needs to be URL encoded, if you ever want to track an upload you'd have to reverse the URL into a key name (aka S3 filename/filepath). 

r?

**Comment by rhelmer** [comment]
One thing we missed - some partners upload symbols that are private and these should go to a separate bucket. For simplicity, we can have this tied to the account (e.g. "partnerA always uploads to private symbol bucket"), or it could be something configurable by the uploader as an optional param in the URL perhaps.

**Comment by peterbe** [comment]
So, how do we decide this? 
Will a certain partner sometimes upload public and sometimes upload private symbols? If so she'll need an option to override in her uploads. 
Otherwise it might be easier to attach a property to the user that uploads always go into a certain bucket.

**Comment by rhelmer** [comment]
@peterbe currently our only use case is that one partner is always private and the rest are always public, so let's just have the ability to tie an account to a bucket if that's simpler.

**Comment by peterbe** [comment]
Should we hardcode that user email to a specific bucket name? in `settings/base.py`.
Or do we need this to be "bigger". I.e. a new UserProfile module with an admin functionality to edit these etc.

**Comment by rhelmer** [comment]
@peterbe I am OK with hardcoding in config for now, maybe a dict to map users to buckets, with a default? I think in the grand scheme of things we don't want to be hosting private symbols for folks unless we really really need to.

**Comment by peterbe** [comment]
@rhelmer e.g. https://gist.github.com/peterbe/dd685444587068081f42
This means this email will never be able to upload public symbols. 
cc @luser

**Comment by luser** [comment]
That looks fine. I think that's exactly what we want--the ability to set a user as "all uploads are private".

**Comment by peterbe** [comment]
Rebased up so hopefully mergeable again.

File: webapp-django/crashstats/symbols/utils.py [hunk 2]
```diff
@@ -5,36 +5,70 @@
...
 
     elif content_type == 'application/x-tar':
         zf = tarfile.TarFile(fileobj=file_object)
         for member in zf.getmembers():
-            if member.isfile():
-                print >>out, member.name.ljust(70),
-                print >>out, str(member.size).rjust(9)
+            # Sometimes when you make a tar file you get a
+            # smaller index file copy that start with "./._".
+            if member.isfile() and not member.name.startswith('./._'):
```
Comment (rhelmer) [inline webapp-django/crashstats/symbols/utils.py]: Hm this is new to me.. looking it up I guess it's an OS X thing, "AppleDouble"?

File: docs/configuring-socorro.rst [hunk 3]
```diff
@@ -152,3 +152,58 @@ All of these jobs can be enabled or disabled in crontabber configuration or by m
 DEFAULT_JOBS in:
 
 https://github.com/mozilla/socorro/blob/master/socorro/cron/crontabber_app.py
+
+
+Symbols S3 uploads
+------------------
```
Comment (rhelmer) [inline docs/configuring-socorro.rst]: Nice, thanks for documenting this :+1: "how do I upload symbols" is a super common question

File: webapp-django/crashstats/symbols/views.py [hunk 4]
```diff
@@ -48,6 +49,80 @@ def check_symbols_archive_content(content, header_length=2):
...
+            settings.SYMBOLS_FILE_PREFIX, member.name
+        )
+        key = bucket.get_key(key_name)
+
+        # let's assume first that we need to add a new key
+        prefix = "+"
+        if key:
+            # key already exists, but is it the same size?
+            if key.size != member.size:
+                # we're changing our minds, so it get uploaded again
```
Comment (rhelmer) [inline webapp-django/crashstats/symbols/views.py]: this "changing our minds" comment is confusing.. maybe just ```# file size in S3 is different, upload the new one```?
If you disagree at least ```s/get/gets/``` :)

**Comment by rhelmer** [comment]
r+ with a silly nit about a comment

**Comment by peterbe** [comment]
rebased and squashed. awaiting travis.

**Pull request merged**